RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,296 @@
+
+The PDK templating engine originally started from [modulesync](https://github.com/voxpupuli/modulesync) which had a fairly simple use case: To generate the configuration files for the Puppet Labs supported modules.
+However since then the PDK use case has greatly expanded in scope. Catering for many CI services, creating files on convert, as well, as keeping files up to date.
+
+The templating system needs to be reviewed, and the architecture changed to support the ever growing need for end user scenarios.
+
+## Background & Assumptions
+
+> What is the context that someone needs to understand the goals of this proposal? This information should be objective
+> facts (i.e. not subjective opinions) about the current state of things.
+
+Evidence that the current templating system is in need of attention
+
+* The number of combinations of settings is continually growing. This is evident in the number of commits, and the growth of the pdk-templates repository
+
+* The complexity of managing the templates is becoming difficult. For example [Jira PDK-1463](https://tickets.puppetlabs.com/browse/PDK-1463?focusedCommentId=680851&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-680851)
+
+* The number of outstanding issues indicates either there are many errors (probably not the case), or that fixing things is difficult (which is more likely)
+
+* [Github Issue 220](https://github.com/puppetlabs/pdk-templates/issues/220) shows that the pdk-templates lack sufficient documentation and testing practices to make it easier for the community to contribute changes, or to have their own template repository.
+
+## Motivation & Goals
+
+### Objectives
+
+**Template Users**
+
+* Template users should be able to compose their required environment from the provided templates instead of opting out of a "kitchen sink" template.
+
+* Template users should be able to see information about all of the available templates available and their settings.
+
+* Template users should be able to verify that their custom template settings are sane and valid
+
+* Template users should be to easily test that their custom template settings work as they expect.
+
+**Template Authors**
+
+* Template authors should be able to easily add/modify templates with clear and documented standards e.g. file locations and good practices
+
+* Template authors should be able to easily generate document from their templates instead of manually keeping the code and documentation in sync.
+
+* Template authors should be able to easily test their templates work as expected, both manually and via automation, like rspec.
+
+* Template authors should be able to advertise what their templates do
+
+* Template authors should be able to use ERB for simple templates, but also put more advanced code into Ruby
+
+### Backwards compatibility
+
+The proposed changes must either be backwards compatible, or be segregated so that Template Users can progressively opt in to version 2.0
```

Comment:
Complementary to what I said earlier about alternatives to `.sync.yml`, we should definitely still support some sort of easy migration path from `.sync.yml` into the new settings.

Reply:
Right now I can't guarantee that V2 would be an easy drop in but I definitely see your point about making the adoption as simple as possible.  Ideally we would add a conversion "script" to transform a V1 sync.yml to V2.  Ideally this would be part of the `pdk update/convert` command.  So the next PDK update would auto-update (Ideally)